lib/index.js

```diff
@@ -57,24 +57,27 @@ internals.Client.prototype.defaults = function (options) {
 };
 
 
+// baseUrl needs to end in a trailing / if it contains paths that need to be preserved
 internals.resolveUrl = function (baseUrl, path) {
 
     if (!path) {
         return baseUrl;
     }
 
-    const parsedPath = Url.parse(path);
-    if (parsedPath.host) {
-        Hoek.assert(parsedPath.protocol, 'Invalid destination path missing protocol');
-        return Url.format(parsedPath);
+    // The URL constructor does a resolve, so need to trim the leading / from path
+    if (path.startsWith('/')) {
+        path = path.substr(1);
     }
 
-    const parsedBase = Url.parse(baseUrl);
-    parsedBase.pathname = parsedBase.pathname + parsedPath.pathname;
-    parsedBase.pathname = parsedBase.pathname.replace(/[/]{2,}/g, '/');
-    parsedBase.search = parsedPath.search;      // Always use the querystring from the path argument
+    const base = new Url.URL(baseUrl);
+    // When the pathname doesn't end in a trailing / it won't be part of the formatted url returned
+    if (!base.pathname.endsWith('/')) {
+        base.pathname += '/';
```

Comment:
FYI, this auto-slashing is not how I would expect it to work. All other functions that resolve against a baseUrl expects a dir path (with a trailing `/`), _or a file path_.

Since this will be a breaking change, I would suggest to drop this behavior.

Reply:
These auto-slashing code paths are meant to ensure compatibility with the way wreck currently works. `new URL` handles these slashes a little differently than I would have expected:

```
> new Url.URL('/bar', 'http://test.com/foo')
URL {
  href: 'http://test.com/bar',
```

```
new Url.URL('bar', 'http://test.com/foo')
URL {
  href: 'http://test.com/bar',
```

```
new Url.URL('bar', 'http://test.com/foo/')
URL {
  href: 'http://test.com/foo/bar',
```

```
new Url.URL('/bar', 'http://test.com/foo/')
URL {
  href: 'http://test.com/bar',
```

I wasn't sure on this particular case if we should document that you need a trailing slash or simply add it. Since this is only for the `baseUrl` it wouldn't make sense that a person would provide a baseUrl of `https://test.com/foo` and not expect `foo` to be part of the `baseUrl`.

Reply:
> Since this is only for the baseUrl it wouldn't make sense that a person would provide a baseUrl of https://test.com/foo and not expect foo to be part of the baseUrl.

Yes it it would, that is how baseUrl normally works.

To fetch resources in a web page, I would use the page url as `baseUrl` (eg. `http://server.com/content/doc.html`) to handle relative references. This will not work with wreck, and I would have to resolve the dir path first before passing it to the option (eg. `{ baseUrl: new URL('.', pageUrl).href` }`).

Reply:
I'm referring to the existing behavior of wreck:

```
Wreck.request('get', '/bar?test=hello', { baseUrl: 'http://localhost:0/foo' })
// req.path is '/foo/bar?test=hello'
```

Are you using wreck in the way that you are describing in your example?

Reply:
Yes, I'm aware that this is the behaviour of wreck. I'm just suggesting that it is changed as well since you are already making a breaking change.

Reply:
I do use wreck in a way where I need to resolve against a baseUrl. However, I'm resolving the urls myself before passing it to wreck.

Reply:
@kanongil I can do that, I'll switch this to breaking.